### ISSUE TITLE
OIDC over Office365/Azure AD - email field empty, user onboarding fails

### ISSUE THREAD

**Issue by kurbar** [goharbor/harbor#13895]
Harbor doesn’t seem to be able to fill in the email field in the database (empty value) so any other attempt fails with the conflict error: `user with same username or email has already been onboarded. .`

Running Harbor version 2.0.2
OIDC is connected to Azure AD (Office365).
Endpoint is set as `https://sts.windows.net/***`
Scopes are set as `openid,offline_access,profile,email`

Checking the core.log file I can see some lines pop up:
```
[WARNING] [/common/utils/oidc/helper.go:331]: Unable to get groups from claims, claims: map[amr:[pwd] aud:*** exp:1.609321256e+09 family_name:*** given_name:*** iat:1.609317356e+09 ipaddr:*** iss:https://sts.windows.net/*** name:*** nbf:1.609317356e+09 oid:*** rh:*** sub:*** tid:*** unique_name:*** upn:*** uti:*** ver:1.0], groups claims key:
[WARNING] [/server/middleware/security/idtoken.go:45][requestID="916096db-c2f7-4314-898d-a76cb302ed3e"]: failed to verify token: oidc: malformed jwt: square/go-jose: compact JWS format must have three parts
[ERROR] [/common/dao/oidc_user.go:135]: fail to insert user, pq: duplicate key value violates unique constraint "harbor_user_email_key"
```

I’ve found numerous issues on GH about this empty email field when using OIDC but none of the solutions seemed to work there. Is it a specific Azure AD issue?

**Comment by reasonerjt** [comment]
The Email is fetched from the userinfo endpoint.
Could you check if the userinfo endpoint is returning `email` as part of the response?

For more details to help you debug see:
https://openid.net/specs/openid-connect-core-1_0.html#UserInfo

**Comment by kurbar** [comment]
@reasonerjt I checked the userinfo endpoint (https://graph.microsoft.com/oidc/userinfo) and the `email` is returned and exists.

```
{
  "sub": "Zk0***",
  "name": "K**** ******",
  "family_name": "*****",
  "given_name": "K****",
  "picture": "https://graph.microsoft.com/v1.0/me/photo/$value",
  "email": "K***@*****.ee"
}
```

**Comment by reasonerjt** [comment]
could you check the user table in `harbor-db` to check if there is a record with the same Email?

**Comment by kurbar** [comment]
Harbor inserts users with empty string e-mail addresses. The first user gets onboarded but any other onboarding fails since an "empty" e-mail address already exists in the users table, giving the duplicate key value. If I manually change the previous users email in the database, then the next one can onboard but after that the same thing continues.

**Comment by reasonerjt** [comment]
I briefly checked the code:
https://github.com/goharbor/harbor/blob/v2.0.2/src/common/utils/oidc/helper.go#L248
https://github.com/goharbor/harbor/blob/v2.0.2/src/common/utils/oidc/helper.go#L273

It seems the Email will be populated based on the response of userinfo, so it's not likely the email remain empty when user is onboarded. 

Given I can't access your env it's hard to determine the root cause.

Could you check if the userinfo endpoint is exposed via the `well-known` endpoint and Harbor DID call this endpoint while user is onboarded?

**Comment by kurbar** [comment]
@reasonerjt Sorry for late reply. I finally got to debugging it further.

I switched Harbor to debug mode and started to onboard a test AD user.

The relevant log lines from `core.log`:
```
Jan 19 18:09:23 ************ core[994]: 2021-01-19T16:09:23Z [DEBUG] [/common/config/store/driver/db.go:43]: failed to get metadata, key:http_authproxy_always_onboard, error:<nil>, skip to load item
Jan 19 18:09:23 ************ core[994]: 2021-01-19T16:09:23Z [DEBUG] [/common/config/store/driver/db.go:43]: failed to get metadata, key:http_authproxy_case_sensitive, error:<nil>, skip to load item
Jan 19 18:09:23 ************ core[994]: 2021-01-19T16:09:23Z [DEBUG] [/common/config/manager.go:192]: failed to get key oidc_groups_claim, error: the configure value is not set
Jan 19 18:09:26 ************ core[994]: 2021-01-19T16:09:26Z [DEBUG] [/common/config/store/driver/db.go:43]: failed to get metadata, key:http_authproxy_always_onboard, error:<nil>, skip to load item
Jan 19 18:09:26 ************ core[994]: 2021-01-19T16:09:26Z [DEBUG] [/common/config/store/driver/db.go:43]: failed to get metadata, key:http_authproxy_case_sensitive, error:<nil>, skip to load item
Jan 19 18:09:26 ************ core[994]: 2021-01-19T16:09:26Z [DEBUG] [/common/config/manager.go:192]: failed to get key oidc_groups_claim, error: the configure value is not set
Jan 19 18:09:27 ************ core[994]: 2021-01-19T16:09:27Z [DEBUG] [/server/middleware/log/log.go:30]: attach request id 145ee112-b6d7-4ba6-9a23-37d97d6a07d2 to the logger for the request POST /c/oidc/onboard
Jan 19 18:09:27 ************ core[994]: 2021-01-19T16:09:27Z [DEBUG] [/common/config/store/driver/db.go:43]: failed to get metadata, key:http_authproxy_always_onboard, error:<nil>, skip to load item
Jan 19 18:09:27 ************ core[994]: 2021-01-19T16:09:27Z [DEBUG] [/common/config/store/driver/db.go:43]: failed to get metadata, key:http_authproxy_case_sensitive, error:<nil>, skip to load item
Jan 19 18:09:27 ************ core[994]: 2021-01-19T16:09:27Z [DEBUG] [/server/middleware/security/unauthorized.go:29][requestID="145ee112-b6d7-4ba6-9a23-37d97d6a07d2"]: an unauthorized security context generated for request POST /c/oidc/onboard
Jan 19 18:09:27 ************ core[994]: 2021-01-19T16:09:27Z [DEBUG] [/common/config/store/driver/db.go:43]: failed to get metadata, key:http_authproxy_always_onboard, error:<nil>, skip to load item
Jan 19 18:09:27 ************ core[994]: 2021-01-19T16:09:27Z [DEBUG] [/common/config/store/driver/db.go:43]: failed to get metadata, key:http_authproxy_case_sensitive, error:<nil>, skip to load item
Jan 19 18:09:27 ************ core[994]: 2021-01-19T16:09:27Z [DEBUG] [/core/controllers/oidc.go:173]: User info string: {"iss":"https://sts.windows.net/***/“,”sub":"Gm*********”,”name”:”S***** T*******”,”email":"","groups":[]}
```


I digged on a bit deeper and I think I have figured out what is happening. I'm not familiar with Golang but with some quick API ref googling and viewing the source:
1. The user info string is read from the session here - https://github.com/goharbor/harbor/blob/v2.0.2/src/core/controllers/oidc.go#L168
2. UserInfo is pulled from AD, user info string is unmarshaled ontop of AD UserInfo - https://github.com/goharbor/harbor/blob/v2.0.2/src/core/controllers/oidc.go#L184-L185

Now if I read `json.Unmarshal` API ref correctly, the 2nd argument indicates what pointer the data is unmarshaled onto. Now if AD UserInfo returned the email, but userInfoStr has email as an empty string, that evaluates that the final email is infact inserted as an empty string since userInfoStr is unmarshaled ontop of AD UserInfo data? (not familiar with Golang, correct me if I'm wrong :)).

**Comment by kurbar** [comment]
@reasonerjt I upgraded to v2.1.3 and the same issue remains. Also tested with automatic onboarding. In both cases the email field in the database is left empty.

**Comment by reasonerjt** [comment]
@kurbar 
The userinfo was generated here, by consolidating the data from userinfo and ID token:
https://github.com/goharbor/harbor/blob/v2.0.2/src/core/controllers/oidc.go#L103

userInfoStr was stored into session:
https://github.com/goharbor/harbor/blob/v2.0.2/src/core/controllers/oidc.go#L126

Was the userinfo endpoint exposed correct in the `openid-configuration` endpoint?

**Comment by kurbar** [comment]
`https://login.microsoftonline.com/{tenant}/v2.0/.well-known/openid-configuration` does give the userinfo endpoint:
```
{
  ...
  "userinfo_endpoint": "https://graph.microsoft.com/oidc/userinfo",
  ...
}
```

**Comment by reasonerjt** [comment]
@kurbar 
You mentioned the endpoint was set as `https://sts.windows.net/***` is there an openid-configuration under this domain and what're the relevant fields?

**Comment by kurbar** [comment]
@reasonerjt sts domain also has an openid-configuration endpoint
`https://sts.windows.net/{tenant}/v2.0/.well-known/openid-configuration`
```
{
  ...
  "userinfo_endpoint": "https://graph.microsoft.com/oidc/userinfo",
  ...
}
```

**Comment by kurbar** [comment]
@reasonerjt I finally managed to figure the issue out.
There are actually two Azure openid-configuration endpoints
```
https://sts.windows.net/{tenant}/.well-known/openid-configuration
https://sts.windows.net/{tenant}/v2.0/.well-known/openid-configuration
```

Harbor was configured to use the first one. And that in fact does not have a `userinfo_endpoint` exposed. I switched it over to the other endpoint and updated `oidc_user.subiss` to reflect it and everything works :).

However, offtopic regarding issue, is the claim where the email is read configurable? `email` exposes and internal Azure email but I'd like to use a different claim which exposes a more human-friendly email.
